packages/remix-dev/compiler/plugins/vanillaExtractPlugin.ts

```diff
@@ -0,0 +1,184 @@
+          let resolveDir = dirname(join(rootDirectory, fileName));
+
+          return {
+            contents: source,
+            loader: "css",
+            resolveDir,
+          };
+        }
+      );
+
+      build.onLoad({ filter: cssFileFilter }, async ({ path: filePath }) => {
+        let identOption: IdentifierOption =
+          mode === "production" ? "short" : "debug";
+
+        let { outputFiles } = await esbuild.build({
+          entryPoints: [filePath],
+          outdir: config.assetsBuildDirectory,
+          assetNames: build.initialOptions.assetNames,
+          bundle: true,
+          external: ["@vanilla-extract"],
+          platform: "node",
+          write: false,
+          plugins: [
+            vanillaExtractTransformPlugin({ rootDirectory, identOption }),
+          ],
+          loader: loaders,
+          absWorkingDir: rootDirectory,
+          publicPath: config.publicPath,
+        });
+
+        let source = outputFiles.find((file) =>
+          file.path.endsWith(".js")
+        )?.text;
+
+        if (!source) {
+          return null;
+        }
+
+        let [contents] = await Promise.all([
+          processVanillaFile({
+            source,
+            filePath,
+            outputCss,
+            identOption,
+          }),
+          outputCss && writeAssets(outputFiles),
+        ]);
+
+        return {
+          contents,
+          resolveDir: dirname(filePath),
+          loader: "js",
+        };
+      });
+    },
+  };
+}
+
+async function writeAssets(
+  outputFiles: Array<esbuild.OutputFile>
+): Promise<void> {
+  await Promise.all(
+    outputFiles
+      .filter((file) => !file.path.endsWith(".js"))
+      .map(async (file) => {
+        await fse.ensureDir(dirname(file.path));
+        await fse.writeFile(file.path, file.contents);
+      })
+  );
+}
+
+const loaderForExtension: Record<string, esbuild.Loader> = {
+  ".js": "js",
+  ".jsx": "jsx",
+  ".ts": "ts",
+  ".tsx": "tsx",
+};
+
+/**
+ * This plugin is used within the child compilation. It applies the Vanilla
+ * Extract file transform to all .css.ts/js files. This is used to add "file
+ * scope" annotations, which is done via function calls at the beginning and end
+ * of each file so that we can tell which CSS file the styles belong to when
+ * evaluating the JS. It's also done to automatically apply debug IDs.
+ */
+function vanillaExtractTransformPlugin({
+  rootDirectory,
+  identOption,
+}: {
+  identOption: IdentifierOption;
+  rootDirectory: string;
+}): esbuild.Plugin {
+  return {
+    name: "vanilla-extract-transform-plugin",
+    setup(build) {
+      build.onLoad({ filter: cssFileFilter }, async ({ path }) => {
+        let source = await fse.readFile(path, "utf-8");
+
+        let contents = await transform({
+          source,
+          filePath: path,
+          rootPath: rootDirectory,
+          packageName: getPackageName(rootDirectory),
```

Comment:
Curious -- what is `packageName` used for in Vanilla Extract? Is it to prevent hash clashes between libs in a monorepo?

Reply:
Good question — it's used to further scope hashes so that you can build library styles ahead of time without clashing, even if the source directory happens to be the same. In our case we probably could just hard code it to "remix-app".